apis/src/responses/challenge.rs

```diff
@@ -70,5 +72,61 @@ impl ChallengeResponse {
             band_lower: challenge.band_lower,
         })
     }
+
+}
+}
+}
+
+fn is_compatible(challenge: &ChallengeResponse, user: &str, details: &ChallengeDetails) -> bool {
+    details.game_type == GameType::from_str(&challenge.game_type).unwrap()
+        && details.rated == challenge.rated
+        && details.time_mode == challenge.time_mode
+        && details.time_base == challenge.time_base
+        && details.time_increment == challenge.time_increment
+        && match details.color_choice {
+            ColorChoice::Random => challenge.color_choice == ColorChoice::Random,
+            ColorChoice::White => challenge.color_choice == ColorChoice::Black,
+            ColorChoice::Black => challenge.color_choice == ColorChoice::White,
+        }
+        && if let Some(opponent) = &details.opponent {
+            opponent == &challenge.challenger.username
+        } else if let Some(opponent) = &challenge.opponent {
+            opponent.username == user
+        } else {
+            challenge.challenger.username != user
+        }
+}
+
+fn has_same_details(challenge: &ChallengeResponse, user: &str, details: &ChallengeDetails) -> bool {
+    let challenge_opponent = challenge
+        .opponent
+        .as_ref()
+        .map(|opponent| opponent.username.as_str());
+
+    details.game_type == GameType::from_str(&challenge.game_type).unwrap()
+        && details.rated == challenge.rated
+        && details.band_lower == challenge.band_upper
+        && details.band_upper == challenge.band_lower
+        && details.time_mode == challenge.time_mode
+        && details.time_base == challenge.time_base
+        && details.time_increment == challenge.time_increment
+        && details.color_choice == challenge.color_choice
+        && challenge_opponent == details.opponent.as_deref()
+        && challenge.challenger.username == user
+}
+
+pub fn create_challenge_handler(
+    user: String,
+    details: ChallengeDetails,
+    mut challenges: Values<String, ChallengeResponse>,
+) -> Option<ChallengeAction> {
+    if details.time_mode == TimeMode::RealTime
```

Comment:
I would have done an early return, but that's fine too.
Just in case I don't make sense:
```rs
if ... {
return None;
}
if let ... {
return Some(...);
}
Some(...)
```